hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionRegistryRpcStubHolder.java

```diff
@@ -0,0 +1,212 @@
+  }
+
+  private void createStubsAndComplete(String clusterId,
+    CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> future) {
+    RpcClient c = RpcClientFactory.createClient(conf, clusterId);
+    ImmutableMap<ServerName, ClientMetaService.Interface> m = createStubs(c, bootstrapNodes);
+    rpcClient = c;
+    addr2Stub = m;
+    future.complete(m);
+  }
+
+  private void createStubs(List<ServerName> bootstrapServers, int index,
+    CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> future) {
+    // use null cluster id here as we do not know the cluster id yet, we will fetch it through this
+    // rpc client and use it to create the final rpc client
+    RpcClient getConnectionRegistryRpcClient = RpcClientFactory.createClient(noAuthConf, null);
+    // user and rpcTimeout are both not important here, as we will not actually send any rpc calls
+    // out, only a preamble connection header, but if we pass null as user, there will be NPE in
+    // some code paths...
+    RpcChannel channel =
+      getConnectionRegistryRpcClient.createRpcChannel(bootstrapServers.get(index), user, 0);
+    ConnectionRegistryService.Interface stub = ConnectionRegistryService.newStub(channel);
+    HBaseRpcController controller = rpcControllerFactory.newController();
+    stub.getConnectionRegistry(controller, GetConnectionRegistryRequest.getDefaultInstance(),
+      new RpcCallback<GetConnectionRegistryResponse>() {
+
+        @Override
+        public void run(GetConnectionRegistryResponse resp) {
+          synchronized (ConnectionRegistryRpcStubHolder.this) {
+            addr2StubFuture = null;
+            if (controller.failed()) {
+              if (ConnectionUtils.isUnexpectedPreambleHeaderException(controller.getFailed())) {
+                // this means we have connected to an old server where it does not support passing
+                // cluster id through preamble connnection header, so we fallback to use null
+                // cluster id, which is the old behavior
+                LOG.debug("Failed to get connection registry info, should be an old server,"
+                  + " fallback to use null cluster id", controller.getFailed());
+                createStubsAndComplete(null, future);
+              } else {
+                LOG.debug("Failed to get connection registry info", controller.getFailed());
+                if (index == bootstrapServers.size() - 1) {
+                  future.completeExceptionally(controller.getFailed());
+                } else {
+                  // try next bootstrap server
+                  createStubs(bootstrapServers, index + 1, future);
+                }
+              }
+            } else {
+              LOG.debug("Got connection registry info: {}", resp);
+              String clusterId = resp.getClusterId();
+              createStubsAndComplete(clusterId, future);
+            }
+          }
+          getConnectionRegistryRpcClient.close();
```

Comment:
do you want to do this in a `finally` block? i think some of the above methods could throw an exception

Reply:
They should not throw any exceptions unless there are code bugs, as there is no throws in the method declaration.
Anyway, let me see if we should always close getConnectionRegistryRpcClient here.

Reply:
After refactoring, I close the rpc client in the newly introduce ClusterIdFetcher class, after finishing all the rpc requests.